cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -404,174 +409,231 @@ func (m *McmManager) GetMachineDeploymentSize(machinedeployment *MachineDeployme
+						case <-time.After(m.conflictRetryInterval):
+							klog.Warningf("could not reset priority annotation for machine %s; Error: %+v, will retry the operation", clone.Name, err)
+							continue
+						}
+					}
+					break
+				}
 			}
+			return nil
+		}()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-			if mclone.Annotations != nil {
-				if mclone.Annotations[machinePriorityAnnotation] == "1" {
-					klog.Infof("Machine %q priority is already set to 1, hence skipping the update", machine.Name)
+// prioritizeMachinesForDeletion prioritizes the targeted machines by updating their priority annotation to 1
+func (m *McmManager) prioritizeMachinesForDeletion(targetMachineRefs []*Ref, mdName string) ([]machineNameNodeNamePair, error) {
+	var expectedToTerminateMachineNodePairs []machineNameNodeNamePair
+	for _, machineRef := range targetMachineRefs {
+		// Trying to update the priority of machineRef till m.maxRetryTimeout
+		ctx, cancelFn := context.WithDeadline(context.Background(), time.Now().Add(m.maxRetryTimeout))
+		err := func() error {
+			defer cancelFn()
+			for {
+				mc, err := m.machineLister.Machines(m.namespace).Get(machineRef.Name)
+				if err != nil {
+					klog.Errorf("Unable to fetch Machine object %s, Error: %+v", machineRef.Name, err)
+					return err
+				}
+				if isMachineTerminating(mc) {
 					break
 				}
-				mclone.Annotations[machinePriorityAnnotation] = "1"
-			} else {
-				mclone.Annotations = make(map[string]string)
-				mclone.Annotations[machinePriorityAnnotation] = "1"
+				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mc.Name, mc.Labels["node"]})
+				if err = m.updateAnnotationOnMachineBeforeDeadline(ctx, mc.Name, machinePriorityAnnotation, "1"); err != nil {
+					return err
+				}
+				// Break out of loop when update succeeds
+				klog.Infof("Machine %s of machineDeployment %s marked with priority 1 successfully", machineRef.Name, mdName)
+				break
 			}
+			return nil
+		}()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return expectedToTerminateMachineNodePairs, nil
+}
 
-			_, err = m.machineClient.Machines(machine.Namespace).Update(context.TODO(), mclone, metav1.UpdateOptions{})
-			if err != nil && time.Now().Before(retryDeadline) {
-				klog.Warningf("Unable to update Machine object %s, Error: %s , will retry in %s", machine.Name, err, conflictRetryInterval)
-				time.Sleep(conflictRetryInterval)
-				continue
-			} else if err != nil {
-				// Timeout occurred
-				klog.Errorf("Unable to update Machine object %s, Error: %s", machine.Name, err)
+// updateAnnotationOnMachineBeforeDeadline returns error only when updating the annotations on machine has been failing consequently and deadline is crossed
+func (m *McmManager) updateAnnotationOnMachineBeforeDeadline(ctx context.Context, mcName string, key, val string) error {
+	for {
+		machine, err := m.machineLister.Machines(m.namespace).Get(mcName)
+		if err != nil {
```

Comment:
again the same thing - if there is an error (other than 404 NotFound) should you not retry?